src/viewport.tsx

```diff
@@ -0,0 +1,121 @@
+import React, {
+  ReactNode,
+  useState,
+  useCallback,
+  useRef,
+  useEffect
+} from 'react'
+import { DebouncedFunc } from 'lodash'
+import lodashThrottle from 'lodash/throttle'
+import { BREAKPOINTS } from './utils'
+
+interface Props {
+  children: ReactNode
+  throttle?: number
+}
+
+const HEADLESS_VIEWPORT_SIZE = { width: 0, height: 0 }
+const VIEWPORT_SIZE_BASE = {
+  breakpoints: BREAKPOINTS,
+  ...getCurrentWindowSize()
+}
+
+const isBrowser = typeof window !== 'undefined'
+
+const ViewportContext = React.createContext(VIEWPORT_SIZE_BASE)
+
+function getCurrentWindowSize() {
+  return isBrowser
+    ? { width: window.innerWidth, height: window.innerHeight }
+    : HEADLESS_VIEWPORT_SIZE
+}
+
+export function ViewportProvider({ throttle = 100, children }: Props) {
+  const [viewportSize, setViewportSize] = useState(VIEWPORT_SIZE_BASE)
+  const throttleHandler = useRef<DebouncedFunc<() => void>>()
+
+  const updateWindowSize = useCallback(() => {
+    setViewportSize({
+      ...VIEWPORT_SIZE_BASE,
+      ...getCurrentWindowSize()
+    })
+  }, [])
+
+  const resizeStart = useCallback((): void => {
+    throttleHandler.current = lodashThrottle(updateWindowSize, throttle)
+
+    updateWindowSize()
+
+    if (isBrowser) {
+      window.addEventListener('resize', throttleHandler.current)
+    }
+  }, [throttle, updateWindowSize])
+
+  const resizeStop = useCallback((): void => {
+    if (!throttleHandler.current) {
+      return
+    }
+
+    updateWindowSize()
```

Comment:
Wasn't sure if we wanted to do this on stop; theoretically we're unmounting (or re-looping) so updating on the stop may not be very useful?

Reply:
I had the same question when I was updating the code, but as it was part of the original logic I didn't want to rock the boat. I have removed just to see and all seems ok so unless someone chimes in with a strong reason, I'll leave it out 👍